### ISSUE TITLE
Can I attach a formatted message to each error variant?

### ISSUE THREAD

**Issue by dan-da** [mcmah309/error_set#3]
I am think something like what thiserror provides.  So each enum variant can have a descriptive message.

I see the warn() method, but that just logs a warning.  I want to attach the error string for caller's usage.

For some apps it can make sense not to use any strings, but I am evaluating error_set for usage in a crate that already uses both thiserror and anyhow to generate errors with descriptive messages.

If it is already possible, please provide example usage.

**Comment by mcmah309** [comment]
Currently, the best way to attach a message would be
```rust
fn example() {
    let var_name_1 = "";
    let var_name_2 = "";
    let x: SetX = SetX::A(AMessage(var_name_1, var_name_2));
}

error_set::error_set! {
    SetX = {
        A(AMessage),
        B
    };
}

#[derive(Debug)]
struct AMessage {
    var_name_1: String,
    var_name_2: String,
}

impl std::error::Error for AMessage {}

impl std::fmt::Display for AMessage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Error A: Need {0} for {1}", self.var_name_1, self.var_name_2)
    }
}
```
I can see how this could not be ideal if you went this route. Logging or referring to the error's documentation is currently the preferred way.

I am open to supporting this though. Something like
```rust
fn example() {
    let var_name_1 = "";
    let var_name_2 = "";
    let x: SetX = SetX::a(var_name_1, var_name_2);
}

error_set::error_set! {
    SetX = {
        A("Error A: Need {var_name_1} for {var_name_2}"),
        B
    };
}
```
Which would work the same as the previous but more concise. Thoughts?

**Comment by dan-da** [comment]
> ```rust
> error_set::error_set! {
>     SetX = {
>         A("Error A: Need {var_name_1} for {var_name_2}"),
>         B
>     };
> }
> ```

I think the way thiserror does it is perhaps more extendable, eg:

```
error_set::error_set! {
    SetX = {
        #[msg("Error A: Need {0} for {1}")]
        A(x, y),
        B
    };
}
```

**Comment by mcmah309** [comment]
I agree. That syntax opens the door for anything that implements `Display` to be valid. Sounds like the best solution so far :+1:

**Comment by mcmah309** [comment]
See https://github.com/mcmah309/error_set/releases/tag/v0.5.0
